### ISSUE TITLE
new extension form without tokens

### ISSUE THREAD

**Issue by SamuraiCrow** [AmosFactory/AmosProfessional#40]
The current token table format has too few slots for extensions.  Replacing the token tables with string hash tables and a parsing expression grammar would be more flexible and extensible.  As a long term goal, I am basing this on my previous experience with Mike Ness on the Mattathias compiler project.

**Comment by SamuraiCrow** [comment]
It occurred to me after I posted that this could be a means to modularity in Amos.library as referenced in #38 and the wasting of slots in #30 for extensions.

**Comment by marc365** [comment]
> "Replacing the token tables with string hash tables"

> "The current token table format has too few slots for extensions."

ok, this work has to be an experiment somewhere in a separate repo as there would be merge conflicts becuase a different developer to me would be doing it, this offers a solution to something I don't consider a problem, I don't know why that would be needed by an amos programmer, I don't understand the scenario here, I support you working on it because I've built my own interpreter and use a similar thing in c#, but until I see the benefit of using it here I can't really help out as I believe the extension system has "enough slots", hopefully it is being planned to integrate transparently when it's finished, and any file format conversions needed would be resolved before hand? so I'm still open to suggestions here, but the way i see it, the use of tokens is an optimization.

**Comment by SamuraiCrow** [comment]
The basic idea is to use the address of a subset parse tree as the token internally so that the language can have as many tokens as nodes in the parse tree.

Among the advantages is that extensions could define or override operators in addition to functions and commands.  This would allow the core language to become modular as well so that the extensions become the language.

 However, as mentioned in the first post, this won't be needed for some time.

**Comment by marc365** [comment]
and don't get the wrong I actually would like to take part in this research as well, but for me (and myabe limited understanding?) it doen's fit as a module in amos yet, but I would really hope that the Mattathias compiler project could be resurrected as a research project that can feed into the amos compiler - I think Mattathias is very advanced and maybe amos is holding that back, we should definitily have a future discussion about this on the IRC channel sometime...

**Comment by SamuraiCrow** [comment]
Thanks.

The code generation of Mattathias was too big and complicated for a 32 bit Amiga but the parser for it was showing hope.  As for bytecode compilation and execution, those techniques can be recycled from LuaJIT, if needed, instead of the LLVM bitcode we were going to use.

This should be a good skunk-works project for the future!

**Comment by SamuraiCrow** [comment]
After some thought about this implementation and its alternatives, it has occurred to me that there are only 2 implications for using a hash values for tokens:
1. If the hash function generates a 16-bit index it takes the same amount of space in tokenized source.
1. Hash collisions will require comparisons to differentiate between duplicate hash values.

This implies some advantages and work-arounds that may improve or retain existing performance.
* Since hashed tokens are dynamically generated hash values, the slot number becomes irrelevant to the number of extensions to be added and to the token values.
* A thin extension mapper can reintroduce backward compatibility to existing extensions via integer hashing.
* Using a [C Minimal Perfect Hash](http://cmph.sourceforge.net/) and taking care not to run afoul of it's MPL 1.1/LGPL dual license, we could actually shrink the hash values to smaller than 16 bits for better cache functionality and eliminate collisions at the expense of becoming a read-only hash lookup.  (The hash would only need regeneration when an extension is added or removed.)
* May be possible without the PEG parser generation just by improving the existing code.

**Comment by marc365** [comment]
> 32bit!

this is very fast, lol:
```
        public static UInt32 Execute(string read)
        {
            UInt32 hashedValue = 0;
            int i = 0;
            UInt32 multiplier = 1;
            while (i < read.Length)
            {
                hashedValue += read[i] * multiplier;
                multiplier *= 37;
                i += 2;
            }
            return hashedValue;
}
```

**Comment by marc365** [comment]
afer the amospro build it runs Make_Checklist - it actually goes into the source and reads what commands/functions are there, including the extensions - I noticed it doesn't all the extensions yet, so it's not a complete command list - it creates the Instruction_list.Asc file.

**Comment by SamuraiCrow** [comment]
Hmmm...  In C# is "int" 32-bit or 64-bit?  I'll assume 32 since your comment would indicate as such.  Also note that multiplying by 32 bits on anything less than a 68060 takes a lot of clock cycles!  A 68060 is equivalent to a Pentium 1, after all.

**Comment by marc365** [comment]
711 commands?

**Comment by marc365** [comment]
lol, I just took a look at the cmph chd as it was the smallest source, there are a few multiplications in there, the hash above is so basic, there must be someway to accelerate it?

**Comment by SamuraiCrow** [comment]
I was looking at the bottom, FCH, one since it is supposed to be simpler.

**Comment by marc365** [comment]
yup, thats doing some mad stuff and not a multiplication in sight, Cool...

**Comment by marc365** [comment]
found some interesting code in +w.s that may be related - it seems that W.s is injecting a command into the list in a2... I believe disecting how this works will enable something special to happen, I'll look a bit later on as I want to stay focused, but i really like seeing this in there :)

search for WAddRoutine in +w.s and you'll find it - my line numbers are different for the time being...

```
;	Ajoute une routine flush
; ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
WAddFlushRoutine
	move.l	a2,-(sp)
	lea	T_MemFlush(a5),a2
	bsr.s	WAddRoutine
	move.l	(sp)+,a2
	rts
;	Insere une routine dans une liste de routines
; ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
;	A1=	Pointeur sur structure
;	A2=	Liste des routines
;		dc.l	0
;		...routine...
WAddRoutine
	tst.l	(a1)			Deja la?
	bmi.s	.Deja
	move.l	(a2),d0
	lsr.l	#1,d0
	bset	#31,d0
	move.l	d0,(a1)			Branche dans la liste
	move.l	a1,(a2)
.Deja	rts
```
